### PR TITLE
test(deps): NUnit 4.0 Migration

### PR DESCRIPTION
**Pull request by tippmar-nr**
This (ridiculously large) PR covers refactoring of our unit tests to run under NUnit 4.0, as well as other minor Nuget package updates in the unit / container / integration / unbounded test solutions. 

_(My apologies for the size of this PR, but I couldn't find a piecemeal way of completing this work across multiple PRs without leaving the build in a broken state...)_

A number of changes were made in the course of this PR, the most relevant being:
* Installation of the `NUnit.CodeAnalyzers` Nuget package, which is great at pointing out issues with our usage of NUnit. As we make future updates to unit tests, this analyzer will help keep us from making mistakes.
* Updated NUnit to v4.0.1, also applied a number of other minor Nuget updates to the test projects - XUnit, Newtonsoft.Json, etc.
* Migrated all existing NUnit API usages to the ["Constraint" model](https://docs.nunit.org/articles/nunit/writing-tests/assertions/assertion-models/constraint.html), which is more of a "fluent" API. This refactoring done primarily by following the recommendations from the CodeAnalyzers package.
* Added `IDisposable` to several existing interface definitions, to allow exposing the `.Dispose()` method. This was necessary because multiple unit tests failed to dispose certain instances during teardown, as pointed out by NUnit.CodeAnalyzers.
* Removed a couple of tests that NUnit.CodeAnalyzers complained about and which, upon investigation, weren't really doing anything useful anyway. Also removed a couple of test cases that were impossible - like passing `null` as the value to a test method that takes a `bool`.
* Removed the AWS Lambda and OpenTracer unit test projects from the Full_Agent solution, as they're not needed and were causing odd behavior during unit test execution.

All Unit and Integration tests still pass with this update.

### CODE REVIEW

File: tests/Agent/NewRelic.Testing.Assertions/NewRelic.Testing.Assertions.csproj [hunk 1]
```diff
@@ -1,16 +1,16 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>net462;net7.0</TargetFrameworks>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="coverlet.collector" Version="6.0.0">
```
Comment (nrcventura) [inline tests/Agent/NewRelic.Testing.Assertions/NewRelic.Testing.Assertions.csproj:6]: Why was this reference removed?

File: tests/Agent/UnitTests/CompositeTests/AdaptiveSamplerTests.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public void BeforeEachTest()
         public void AfterEachTest()
         {
             _compositeTestAgent.Dispose();
-            _adaptiveSampler = null;
+            _adaptiveSampler.Dispose();
```
Comment (nrcventura) [inline tests/Agent/UnitTests/CompositeTests/AdaptiveSamplerTests.cs:41]: Should we still set this value to null, in addition to the dispose call?

File: tests/Agent/UnitTests/CompositeTests/CrossAgentTests/SecurityPolicies/SecurityPoliciesCrossAgentTests.cs [hunk 3]
```diff
@@ -58,9 +58,10 @@ public void Setup()
         }
 
         [TearDown]
-        public static void TearDown()
+        public void TearDown()
         {
             _compositeTestAgent.Dispose();
```
Comment (nrcventura) [inline tests/Agent/UnitTests/CompositeTests/CrossAgentTests/SecurityPolicies/SecurityPoliciesCrossAgentTests.cs:63]: The _compositeTestAgent field should probably not be static too.

File: tests/Agent/UnitTests/Core.UnitTest/Configuration/DefaultConfigurationTests.cs [hunk 4]
```diff
@@ -1006,6 +1005,7 @@ public string IgnoreErrorsAndIgnoreClassesCombineTests(string[] ignoreClasses, s
             return string.Join(",", _defaultConfig.IgnoreErrorsConfiguration.Keys);
         }
 
+#pragma warning disable NUnit1005 // The type of the value specified via ExpectedResult must match the return type of the method
```
Comment (nrcventura) [inline tests/Agent/UnitTests/Core.UnitTest/Configuration/DefaultConfigurationTests.cs:1008]: Should we submit a tech debt ticket so that we don't need this pragma anymore?

File: tests/Agent/UnitTests/Core.UnitTest/Database/SqlParsingCrossAgentTest.cs [hunk 5]
```diff
@@ -18,18 +18,21 @@ namespace NewRelic.Agent.Core.NewRelic.Agent.Core.Database
...
+                Assert.That(parsed.Operation, Is.EqualTo(expectedOperation.ToLower()), string.Format("Expected operation {0} but was {1}", expectedOperation, parsed.Operation));
+                Assert.That(parsed.Model, Is.EqualTo(expectedTable?.ToLower()), string.Format("Expected table {0} but was {1}", expectedTable, parsed.Model));
+            });
         }
 
 
         // You can uncomment this to measure timing of sql parsing
         //[Test]
-        public void Timing()
+        private void Timing()
```
Comment (nrcventura) [inline tests/Agent/UnitTests/Core.UnitTest/Database/SqlParsingCrossAgentTest.cs:35]: Instead of making this private, should this just be removed? It looks like it's an "test" that was used to time things.

File: tests/Agent/UnitTests/Core.UnitTest/Segments/MessageBrokerSegmentDataTests.cs [hunk 6]
```diff
@@ -18,7 +18,9 @@ public class MessageBrokerSegmentDataTests
     {
         #region IsCombinableWith
 
+#pragma warning disable NUnit1028 // The non-test method is public
```
Comment (nrcventura) [inline tests/Agent/UnitTests/Core.UnitTest/Segments/MessageBrokerSegmentDataTests.cs:21]: Should we submit a tech debt item so that we don't need this pragma?

File: tests/Agent/UnitTests/Core.UnitTest/Segments/MethodSegmentDataTests.cs [hunk 7]
```diff
@@ -17,7 +17,9 @@ public class MethodSegmentDataTests
     {
         #region IsCombinableWith
 
+#pragma warning disable NUnit1028 // The non-test method is public
```
Comment (nrcventura) [inline tests/Agent/UnitTests/Core.UnitTest/Segments/MethodSegmentDataTests.cs:20]: Should we submit a tech debt ticket for this pragma too?

File: tests/Agent/UnitTests/Core.UnitTest/Segments/SimpleSegmentDataTests.cs [hunk 8]
```diff
@@ -20,7 +20,9 @@ public class SimpleSegmentDataTests
     {
         #region IsCombinableWith
 
+#pragma warning disable NUnit1028 // The non-test method is public
```
Comment (nrcventura) [inline tests/Agent/UnitTests/Core.UnitTest/Segments/SimpleSegmentDataTests.cs:23]: Submit a ticket for this too?

File: tests/Agent/UnitTests/Core.UnitTest/Wrapper/AgentWrapperApi/Builders/TransactionTests.cs [hunk 9]
```diff
@@ -73,27 +76,34 @@ public void TransactionDoesNotImplementIDisposable()
...
-            Assert.IsNotInstanceOf<IDisposable>(_transaction);
+            Assert.That(_transaction, Is.Not.InstanceOf<IDisposable>());
         }
 
         [Test]
         public void TransactionFinalizedEvent_IsPublished_IfNotEndedCleanly()
         {
-            Assert.NotNull(_transaction);
+
+#pragma warning disable NUnit2018 // can't use constraint model here, no idea why but test fails if you do
```
Comment (nrcventura) [inline tests/Agent/UnitTests/Core.UnitTest/Wrapper/AgentWrapperApi/Builders/TransactionTests.cs:86]: Should we add a ticket to the backlog for this?

File: tests/Agent/NewRelic.Testing.Assertions/NewRelic.Testing.Assertions.csproj (hunk 1 above)
Comment (tippmar-nr) [reply tests/Agent/NewRelic.Testing.Assertions/NewRelic.Testing.Assertions.csproj:6]: Because this isn't a unit test project; it's a utility project.

File: tests/Agent/UnitTests/CompositeTests/AdaptiveSamplerTests.cs (hunk 2 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/CompositeTests/AdaptiveSamplerTests.cs:41]: I'm not sure, but it probably wouldn't hurt anything.

File: tests/Agent/UnitTests/Core.UnitTest/Configuration/DefaultConfigurationTests.cs (hunk 4 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/Core.UnitTest/Configuration/DefaultConfigurationTests.cs:1008]: I'll just refactor real quick. Not a big deal.

File: tests/Agent/UnitTests/Core.UnitTest/Segments/MessageBrokerSegmentDataTests.cs (hunk 6 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/Core.UnitTest/Segments/MessageBrokerSegmentDataTests.cs:21]: Another easy refactor. Done.

File: tests/Agent/UnitTests/Core.UnitTest/Segments/SimpleSegmentDataTests.cs (hunk 8 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/Core.UnitTest/Segments/SimpleSegmentDataTests.cs:23]: Refactored.

File: tests/Agent/UnitTests/Core.UnitTest/Segments/MethodSegmentDataTests.cs (hunk 7 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/Core.UnitTest/Segments/MethodSegmentDataTests.cs:20]: Refactored

File: tests/Agent/UnitTests/Core.UnitTest/Database/SqlParsingCrossAgentTest.cs (hunk 5 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/Core.UnitTest/Database/SqlParsingCrossAgentTest.cs:35]: Done.

File: tests/Agent/UnitTests/CompositeTests/CrossAgentTests/SecurityPolicies/SecurityPoliciesCrossAgentTests.cs (hunk 3 above)
Comment (tippmar-nr) [reply tests/Agent/UnitTests/CompositeTests/CrossAgentTests/SecurityPolicies/SecurityPoliciesCrossAgentTests.cs:63]: Good catch! Done.

**Review by nrcventura — APPROVED** [review]
